ucb-bar/chipyard
# Initial draft of local FPGA support, targeting Arty

**Type of change**:  new feature


**Impact**: rtl change

**Release Notes**


This is a first draft of local FPGA support, initially targeting the Xilinx ARTY dev board. Most of this initial implementation has been taken from SiFive's Freedom platform. I have created a subdirectory `fpga` where the Chisel, Makefiles, bootrom, and fpga-shells submodule live. fpga-shells has not been forked or modified other than to update the commit from that included with Freedom. I am open to changing the directory structure or naming, since perhaps a top-level `fpga` directory could be confusing when FireSim lives in `generators`.

I have also added initialization to the init scripts.

To test this implementation, run
```
./scripts/init-fpga.sh
```
chdir into fpga, and run:
```
make -f Makefile.e300artydevkit verilog
make -f Makefile.e300artydevkit mcs
```
to generate the Verilog and bitfile. It can then be programmed to an Arty with Vivado and interacted with via the freedom-e-sdk.

To build with a MMIO Chipyard generator like GCD, add its mixin to `fpga/src/main/scala/arty` and change the base address in GCD.scala to something like 0x10017000. The re-run the make steps above. I've written a simple GCD program in freedom-e-sdk that tests the GCD hardware, and it is working.


As mentioned in PR #669, all `vsrcs` should be included with the `addResources` FIRRTL utility.
